Sources/CameraKit/Capture/CameraService.swift

```diff
@@ -5,141 +5,192 @@
+            guard captureSession.canAddOutput(videoDataOutput) else {
+                throw CameraServiceError.initializationError
+            }
+            // cameraCaptureVideoOutput.alwaysDiscardsLateVideoFrames = true
+            videoDataOutput.setSampleBufferDelegate(self, queue: DispatchQueue.global(qos: .default))
+            captureSession.addOutput(videoDataOutput)
+            videoDataOutput.connection(with: .video)?.videoOrientation = .portrait
+            videoDataOutput.connection(with: .video)?.isVideoMirrored = cameraDirection == .front
         }
+        captureSession.commitConfiguration()
+        startSessionIfNeeded()
     }
-    
+
     func restartCaptureSession() {
-        guard !cameraCaptureSession.isRunning else { return }
+        guard !captureSession.isRunning else { return }
         startSessionIfNeeded()
     }
-    
+
     func stopCaptureSessionIfNeeded() {
-        guard cameraCaptureSession != nil else { return }
-        cameraCaptureSession.stopRunning()
+        guard captureSession != nil else { return }
+        captureSession.stopRunning()
     }
-    
+
+    @discardableResult
+    func zoom(to factor: Double) throws -> Double {
+        guard let currentCameraDevice else {
+            throw CameraServiceError.currentCameraNotSet
+        }
+        let factor = min(
+            max(factor, currentCameraDevice.minAvailableVideoZoomFactor),
+            currentCameraDevice.activeFormat.videoMaxZoomFactor
+        )
+        try currentCameraDevice.lockForConfiguration()
+        defer { currentCameraDevice.unlockForConfiguration() }
+        currentCameraDevice.videoZoomFactor = factor
+
+        return factor
+    }
+
     func takePicture() {
         let settings = AVCapturePhotoSettings(format: [AVVideoCodecKey: AVVideoCodecType.jpeg])
-        cameraCapturePhotoOutput.capturePhoto(with: settings, delegate: self)
+        photoOuput.capturePhoto(with: settings, delegate: self)
     }
-    
+
     func switchCamera() throws {
-        cameraDirection = cameraDirection == .front ? .back : .front
-        try setupCameraSession()
+        cameraDirection = switch cameraDirection {
+        case .front:
+                .back
+        case .back:
+                .front
+        }
+        if captureSession.isRunning {
+            guard let currentCameraDeviceInput,
+                  let newCameraDevice = retrieveCamera() else {
+                throw CameraServiceError.currentCameraNotSet
+            }
+            captureSession.beginConfiguration()
+            defer { captureSession.commitConfiguration() }
+            captureSession.removeInput(currentCameraDeviceInput)
+            guard let newCameraDeviceInput = try? AVCaptureDeviceInput(device: newCameraDevice),
+                  captureSession.canAddInput(newCameraDeviceInput) else {
+                throw CameraServiceError.initializationError
+            }
+            captureSession.addInput(newCameraDeviceInput)
```

Comment:
I do believe that we should call `videoDataOutput.connection(with: .video)?.isVideoMirrored = cameraDirection == .front`  here as well. WDYT?